include/revng/Pipeline/Kind.h

```diff
@@ -49,6 +59,41 @@ class Kind : public DynamicHierarchy<Kind> {
     revng_assert(TheRank != nullptr);
   }
 
+  template<RankSpecialization BaseRank, typename... T>
+  requires(RankConvertibleTo<BaseRank, T> and...)
+    Kind(llvm::StringRef Name,
+         const BaseRank &TheRank,
+         LocationReferences<T...> Locations,
+         std::initializer_list<const Kind *> DefinitionKinds,
+         bool SelfDefinition = false) :
+    DynamicHierarchy<Kind>(Name),
+    Register(*this),
+    TheRank(&TheRank),
+    DefinitionKinds(DefinitionKinds) {
+    if (SelfDefinition)
+      this->DefinitionKinds.push_back(this);
+    std::apply([&](auto &&...Loc) { (this->Locations.push_back(&Loc), ...); },
+               Locations.References);
+  }
+
+  template<RankSpecialization BaseRank, typename... T>
+  requires(RankConvertibleTo<BaseRank, T> and...)
+    Kind(llvm::StringRef Name,
+         Kind &Parent,
+         const BaseRank &TheRank,
+         LocationReferences<T...> Locations,
+         std::initializer_list<const Kind *> DefinitionKinds,
+         bool IncludeSelf = false) :
+    DynamicHierarchy<Kind>(Name, Parent),
+    Register(*this),
+    TheRank(&TheRank),
+    DefinitionKinds(DefinitionKinds) {
+    if (IncludeSelf)
+      this->DefinitionKinds.push_back(this);
+    std::apply([&](auto &&...Loc) { (this->Locations.push_back(&Loc), ...); },
+               Locations.References);
+  }
```

Comment:
1. They two boolean arguments were named identically, I guess I had a slip up. For now I guess it can be dropped.
2a. For DefinitionKinds: without the `initializer_list` the braced initializer cannot be used, so since Kinds are static there'd be extra static variables to then pass them as `llvm::ArrayRef`
2b. Ranks *need* to be passed as a `std::tuple<const T &...>` (and the reason for the LocationReferences stuff, but it can be simplified), otherwise you cannot check that a location is a sublocation of the other.
3. I'll take a look but there's a bit of indirection going on with `LLVMKind` which I'd rather avoid touching

Reply:
2. For `DefinitionKinds`: if the argument is a `std::vector` you can't pass an initializer list? o_O
   ```c++
   struct X {
     X(std::vector<int>){}
   };
   X mmm({1,2,3});
   ```
   For `Ranks`: what we get in return for the extra complexity is very very little.
   If you want to keep it, we can rename `LocationReferences` into `ChildrenOfOrSelfRanksTuple<Self, T...>` and push in there the check.
3. Ask help to massimo, if needed.

Reply:
2. `DefinitionKinds`: ok my bad, I was remembering `llvm::ArrayRef` which doesn't accept an initializer list, changed to `std::vector`
  `Ranks`: The template deduction stuff was getting cubersome, removed everything and switched back to `std::tuple` for simplicity
3. Dropped the old constructors

Reply:
2. Why do you need a `tuple` here? I think accepting a `std::vector<Rank *>` would do the job.

Reply:
there's the `RankConvertibleTo<BaseRank, T>` which works on `TypedRank`, so I can't downcast

Reply:
Ah OK, I thought your previous message meant that you were dropping that.
OK then!